R/coord-polar2.R

```diff
@@ -0,0 +1,580 @@
+    )
+  },
+
+  render_axis_h = function(self, panel_params, theme) {
+    if (self$r_axis_inside) {
+      return(list(top = zeroGrob(), bottom = zeroGrob()))
+    }
+    list(
+      top    = panel_guides_grob(panel_params$guides, position = "top",    theme = theme),
+      bottom = panel_guides_grob(panel_params$guides, position = "bottom", theme = theme)
+    )
+  },
+
+  render_bg = function(self, panel_params, theme) {
+
+    bbox  <- panel_params$bbox %||% list(x = c(0, 1), y = c(0, 1))
+    arc   <- panel_params$arc  %||% c(0, 2 * pi)
+    dir   <- self$direction
+    donut <- panel_params$donut
+
+    theta_lim <- panel_params$theta.range
+    theta_maj <- panel_params$theta.major
+    theta_min <- setdiff(panel_params$theta.minor, theta_maj)
+
+    if (length(theta_maj) > 0) {
+      theta_maj <- theta_rescale2(theta_maj, theta_lim, arc, dir)
+    }
+    if (length(theta_min) > 0) {
+      theta_min <- theta_rescale2(theta_min, theta_lim, arc, dir)
+    }
+    theta_fine <- seq(self$arc[1], self$arc[2], length.out = 100)
+
+    r_fine <- r_rescale2(panel_params$r.major, panel_params$r.range,
+                         panel_params$donut)
+
+    # This gets the proper theme element for theta and r grid lines:
+    #   panel.grid.major.x or .y
+    grid_elems <- paste(
+      c("panel.grid.major.", "panel.grid.minor.", "panel.grid.major."),
+      c(self$theta, self$theta, self$r), sep = ""
+    )
+    grid_elems <- lapply(grid_elems, calc_element, theme = theme)
+    majortheta <- paste("panel.grid.major.", self$theta, sep = "")
+    minortheta <- paste("panel.grid.minor.", self$theta, sep = "")
+    majorr     <- paste("panel.grid.major.", self$r,     sep = "")
+
+    bg_element <- calc_element("panel.background", theme)
+    if (!inherits(bg_element, "element_blank")) {
+      background <- data_frame0(
+        x = c(Inf,  Inf, -Inf, -Inf),
+        y = c(Inf, -Inf, -Inf,  Inf)
+      )
+      background <- coord_munch(self, background, panel_params, is_closed = TRUE)
+      bg_gp <- gpar(
+        lwd = len0_null(bg_element$linewidth * .pt),
+        col = bg_element$colour, fill = bg_element$fill,
+        lty = bg_element$linetype
+      )
+      background <- polygonGrob(
+        x = background$x, y = background$y,
+        gp = bg_gp
+      )
+    } else {
+      background <- zeroGrob()
+    }
+
+    ggname("grill", grobTree(
+      background,
+      theta_grid(theta_maj, grid_elems[[1]], donut, bbox),
+      theta_grid(theta_min, grid_elems[[2]], donut, bbox),
+      element_render(
+        theme, majorr, name = "radius",
+        x = rescale(rep(r_fine, each = length(theta_fine)) *
+          rep(sin(theta_fine), length(r_fine)) + 0.5, from = bbox$x),
+        y = rescale(rep(r_fine, each = length(theta_fine)) *
+          rep(cos(theta_fine), length(r_fine)) + 0.5, from = bbox$y),
+        id.lengths = rep(length(theta_fine), length(r_fine)),
+        default.units = "native"
+      )
+    ))
+  },
```

Comment:
Can you point me to where the background gets shaped like the plotting area (circle, partial circle, donut, etc) - I can't seem to find it in here

Reply:
Yeah it is in L294-L298. Essentially, we make an infinitely sized rectangle that we let the coord munching reshape to what it should be.